splitgraph/cloud/tunnel_client.py

```diff
@@ -0,0 +1,94 @@
+import os
+import subprocess
+from os import path
+from typing import Optional
+
+RATHOLE_CLIENT_CONFIG_FILENAME = "rathole-client.toml"
+
+RATHOLE_CLIENT_CONFIG_TEMPLATE = """
+[client]
+remote_addr = "{tunnel_connect_address}"
+
+[client.transport]
+type = "tls"
+
+[client.transport.tls]
+{trusted_root_line}
+hostname = "{tls_hostname}"
+
+[client.services."{section_id}"]
+local_addr = "{local_address}"
+token = "{secret_token}"
+
+"""
+
+
+def get_rathole_client_config(
+    tunnel_connect_address: str,
+    tls_hostname: str,
+    local_address: str,
+    secret_token: str,
+    section_id: str,
+    trusted_root: Optional[str],
+) -> str:
+    trusted_root_line = f'trusted_root = "{trusted_root}"' if trusted_root else ""
+    return RATHOLE_CLIENT_CONFIG_TEMPLATE.format(
+        tunnel_connect_address=tunnel_connect_address,
+        tls_hostname=tls_hostname,
+        local_address=local_address,
+        secret_token=secret_token,
+        section_id=section_id,
+        trusted_root_line=trusted_root_line,
+    )
+
+
+def get_config_dir():
+    from splitgraph.config import CONFIG
+    from splitgraph.config.config import get_singleton
+
+    return os.path.dirname(get_singleton(CONFIG, "SG_CONFIG_FILE"))
+
+
+def get_rathole_client_binary_path():
+    return os.path.join(get_config_dir(), "rathole")
+
+
+def write_rathole_client_config(
+    section_id: str,
+    secret_token: str,
+    tunnel_connect_host: str,
+    tunnel_connect_port: int,
+    local_address: str,
+    tls_hostname: Optional[str],
+) -> str:
+    # in production, this will be None, but for dev instances, we need to
+    # specify rootCA.pem
+    trusted_root = os.environ.get("REQUESTS_CA_BUNDLE") or os.environ.get("SSL_CERT_FILE")
+    rathole_client_config = get_rathole_client_config(
+        tunnel_connect_address=f"{tunnel_connect_host}:{tunnel_connect_port}",
+        tls_hostname=tls_hostname or tunnel_connect_host,
+        local_address=local_address,
+        secret_token=secret_token,
+        section_id=section_id,
+        trusted_root=trusted_root,
+    )
+    config_filename = path.join(get_config_dir(), RATHOLE_CLIENT_CONFIG_FILENAME)
+    with open(config_filename, "w") as f:
+        f.write(rathole_client_config)
+    return config_filename
+
+
+def launch_rathole_client(rathole_client_binary_path: str, rathole_client_config_path: str) -> int:
+    command = [rathole_client_binary_path, "--client", rathole_client_config_path]
+    process = subprocess.Popen(command, stdout=subprocess.PIPE, stderr=subprocess.STDOUT)
+    if process.stdout:
+        while True:
+            output = str(process.stdout.readline())
+            if output == "" and process.poll() is not None:
+                break
+            if output:
+                print(output.strip())
+        rc = process.poll()
+        return rc or 0
+    else:
+        return -1
```

Comment:
I think `subprocess.check_call` will handle everything for you, including forwarding output and Ctrl+C:

```suggestion
    subprocess.check_call(command)
```

This will exit when the process completes and raise a `CalledProcessError` if it exits with a non-zero code, so you don't need to return a -1 here (this can be a `-> None` function)

ex:

```python
import subprocess

subprocess.check_call(["watch", "echo 'test'"])
```

or

```python
subprocess.check_call(["cat", "nonexistent.txt"])
```

```
cat: nonexistent.txt: No such file or directory
Traceback (most recent call last):
  File "/home/mildbyte/.config/JetBrains/PyCharmCE2021.2/scratches/scratch_134.py", line 6, in <module>
    subprocess.check_call(["cat", "nonexistent.txt"])
  File "/home/mildbyte/.pyenv/versions/3.9.4/lib/python3.9/subprocess.py", line 373, in check_call
    raise CalledProcessError(retcode, cmd)
subprocess.CalledProcessError: Command '['cat', 'nonexistent.txt']' returned non-zero exit status 1.
```